llvm/tools/spir-to-ir-wrapper/spir-to-ir-wrapper.cpp

```diff
@@ -0,0 +1,142 @@
+#include "llvm/Support/FileSystem.h"
+#include "llvm/Support/InitLLVM.h"
+#include "llvm/Support/Path.h"
+#include "llvm/Support/Program.h"
+#include "llvm/Support/SourceMgr.h"
+#include "llvm/Support/StringSaver.h"
+
+using namespace llvm;
+
+// InputFilename - The filename to read from.
+static cl::opt<std::string> InputFilename(cl::Positional,
+                                          cl::value_desc("<input spv file>"),
+                                          cl::desc("<input file>"));
+
+// Output - The filename to output to.
+static cl::opt<std::string> Output("o", cl::value_desc("output IR filename"),
+                                   cl::desc("output filename"));
+
+// LlvmSpirvOpts - The filename to output to.
+static cl::opt<std::string>
+    LlvmSpirvOpts("llvm-spirv-opts", cl::value_desc("llvm-spirv options"),
+                  cl::desc("options to pass to llvm-spirv"));
+
+static void error(const Twine &Message) {
+  llvm::errs() << "spir-to-ir-wrapper: " << Message << '\n';
+  exit(1);
+}
+
+// Convert the SPIR-V to LLVM-IR.
+static int convertSPIRVToLLVMIR(const char *Argv0) {
+  // Find llvm-spirv.  It is expected this resides in the same directory
+  // as spir-to-ir-wrapper.
+  StringRef ParentPath = llvm::sys::path::parent_path(Argv0);
+  llvm::ErrorOr<std::string> LlvmSpirvBinary =
+      llvm::sys::findProgramByName("llvm-spirv", ParentPath);
+  if (!LlvmSpirvBinary)
+    LlvmSpirvBinary = llvm::sys::findProgramByName("llvm-spirv");
+
+  SmallVector<StringRef, 6> LlvmSpirvArgs = {"llvm-spirv", "-r", InputFilename,
+                                             "-o", Output};
+
+  // Add any additional options specified by the user.
+  SmallVector<const char *, 8> TargetArgs;
+  llvm::BumpPtrAllocator BPA;
+  llvm::StringSaver S(BPA);
+  if (!LlvmSpirvOpts.empty()) {
+    // Tokenize the string.
+    llvm::cl::TokenizeGNUCommandLine(LlvmSpirvOpts, S, TargetArgs);
+    std::copy(TargetArgs.begin(), TargetArgs.end(),
+              std::back_inserter(LlvmSpirvArgs));
+  }
+
+  return llvm::sys::ExecuteAndWait(LlvmSpirvBinary.get(), LlvmSpirvArgs);
+}
+
+static int copyInputToOutput(void) {
+  // When given an output file, just copy the input to the output
+  if (!Output.empty() && !InputFilename.empty()) {
+    llvm::sys::fs::copy_file(InputFilename, Output);
+  }
+  return 0;
+}
+
+static bool isSPIRVBinary(const std::string &File) {
+  auto FileOrError = MemoryBuffer::getFile(File, /*IsText=*/false,
+                                           /*RequiresNullTerminator=*/false);
+  if (!FileOrError)
+    return false;
+  std::unique_ptr<MemoryBuffer> FileBuffer = std::move(*FileOrError);
+  return SPIRV::isSpirvBinary(FileBuffer->getBuffer().str());
+}
+
+static bool isLLVMIRBinary(const std::string &File) {
+  if (File.size() < sizeof(unsigned))
+    return false;
+
+  StringRef Ext = llvm::sys::path::has_extension(File)
+                      ? llvm::sys::path::extension(File).drop_front()
+                      : "";
```

Comment:
Judging by the [API's description](https://llvm.org/doxygen/namespacellvm_1_1sys_1_1path.html#ad7c38f73fbf2f1a4ed9578621c6bfe8b), this should be a safe way to simplify the code:
```suggestion
  StringRef Ext = llvm::sys::path::extension(File).drop_front();
```

Reply:
It looks like drop_front() requires a size of at least 1, so if extension is empty it will fail.

Reply:
Ah, I've somehow missed the whole point then. Sorry for the fuss.